crates/ecs/src/lib.rs

```diff
@@ -503,74 +499,348 @@ pub struct World {
-            let component_typeid = TypeId::of::<ComponentType>();
-            self.stored_types.push(component_typeid);
-            let archetype_index = 0;
             self.component_typeid_to_archetype_indices
-                .entry(component_typeid)
-                .or_default()
-                .insert(archetype_index);
+                .get_mut(&type_id)
+                .expect("Type ID should exist for previously existing types")
+                .insert(target_archetype_idx);
+        }
+
+        target_archetype.entity_order.push(entity);
+        target_archetype.entity_to_component_index.insert(entity, 0);
+        self.archetypes.push(target_archetype);
+        self.entity_to_archetype_index
+            .insert(entity, target_archetype_idx);
+
+        Ok(target_archetype_idx)
+    }
+
+    fn update_source_archetype_after_entity_move(
+        &mut self,
+        entity: Entity,
+        source_archetype_index: ArchetypeIndex,
+        source_component_vec_idx: ComponentIndex,
+    ) -> WorldResult<()> {
+        let source_archetype = self
+            .archetypes
+            .get_mut(source_archetype_index)
+            .ok_or(WorldError::ArchetypeDoesNotExist(source_archetype_index))?;
+
+        // Update old archetypes component vec index after moving entity
+        if let Some(&swapped_entity) = source_archetype.entity_order.last() {
+            if swapped_entity != entity {
+                source_archetype
+                    .entity_to_component_index
+                    .insert(swapped_entity, source_component_vec_idx);
+            }
+        }
+        // Remove entity component index existing in its old archetype
+        source_archetype
+            .entity_order
+            .swap_remove(source_component_vec_idx);
+
+        source_archetype.entity_to_component_index.remove(&entity);
+        Ok(())
+    }
+
+    fn add_component_to_entity<ComponentType: Debug + Send + Sync + 'static>(
+        &mut self,
+        entity: Entity,
+        component: ComponentType,
+    ) -> WorldResult<()> {
+        let source_archetype_index = *self
+            .entity_to_archetype_index
+            .get(&entity)
+            .ok_or(WorldError::EntityDoesNotExist(entity))?;
+
+        let source_archetype = self
+            .archetypes
+            .get(source_archetype_index)
+            .ok_or(WorldError::ArchetypeDoesNotExist(source_archetype_index))?;
+
+        let source_component_vec_idx = *source_archetype
+            .entity_to_component_index
+            .get(&entity)
+            .expect(
+                "Entity should yield a component index
+             in archetype since the the relevant archetype
+              was fetched from the entity itself.",
+            );
+
+        let (target_archetype_exists, (source_type_ids, _)) =
+            self.find_target_archetype::<ComponentType>(entity, false)?;
+
+        match target_archetype_exists {
```

Comment:
I would have changed this method `find_target_archetype` to  `find_or_create_target_archetype` so that you either find a matching archetype or create a new one.

This way you wont need the match statement and can instead handle both cases in the same way.

You don't have to do this, but I think it would make the code simpler.

Reply:
Yes I agree that it would be more readable however it might introduce a performance reduction. 

Atm I'm creating the new target archetype while moving components. Looping over the component types once while creating the archetype and once again while moving which is what would be required to use a `find_or_create_target_archetype` followed by the actual entity move. This is bypassed by the current implementation.

Choosing not to implementing this now because of this or am I wrong in my assumption regarding performance?